primitives/io/src/tasks.rs

```diff
@@ -0,0 +1,114 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Parallel tasks.
+
+#[cfg(feature = "std")]
+mod inner {
+    use std::sync::mpsc;
+    use sp_externalities::ExternalitiesExt as _;
+    use crate::TaskExecutorExt;
+
+    /// Task handle (wasm).
+    ///
+    /// This can be `join`-ed to get (blocking) the result of
+    /// the spawned task execution.
+    pub struct DataJoinHandle {
+        receiver: mpsc::Receiver<Vec<u8>>,
+    }
+
+    impl DataJoinHandle {
+        /// Join handle returned by `spawn` function
+        pub fn join(self) -> Vec<u8> {
+            self.receiver.recv().expect("Essntial task failure: forked runtime terminated before sending result.")
+        }
+    }
+
+    /// Spawn new runtime task (native).
+    pub fn spawn(entry_point: fn(Vec<u8>) -> Vec<u8>, data: Vec<u8>) -> Result<DataJoinHandle, &'static str> {
+        let scheduler = sp_externalities::with_externalities(|mut ext| ext.extension::<TaskExecutorExt>()
```

Comment:
Should we run sequentially when we do not have a `TaskExecutorExt` extension?
I am a bit hesitant, maybe registering a sequential `TaskExecutorExt` would be best. Seems to be not directly related to this PR (IIRC task executor ext is also used for batch verification, so there it make sense to not allowing it since the batch verification is optional, in our case it is also kinda optional since not all runtimes will be using tasks::spawn so failing seems also ok). So just wondering about it.

Reply:
https://github.com/paritytech/substrate/pull/7089#discussion_r488472607

answered there